### ISSUE TITLE
Making cffi work on Mac M1

### ISSUE THREAD

**Issue by almarklein** [pygfx/wgpu-py#190]
See (the top post of) #136 for the bigger picture. I'm trying to collect what we know here (and may update).

It looks like we have two problems with CFFI.

### Problem

There are no binary wheels provided for macos_arm64, so it tries to use the x86 one (or something like this). Also see the corresponding issue: https://foss.heptapod.net/pypy/cffi/-/issues/506 - a workaround has been proposed, and it looks like they'll wait for someone else to fix it.

This error that you'd get:

> E ImportError: dlopen(/Users/xx/pypoetry/virtualenvs/wgpu-5bSf_T1V-py3.9/lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so, 2): Symbol not found: _ffi_prep_closure
> E Referenced from: /Users/xx/pypoetry/virtualenvs/wgpu-5bSf_T1V-py3.9/lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so
> E Expected in: flat namespace
> E in /Users/xx/pypoetry/virtualenvs/wgpu-5bSf_T1V-py3.9/lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so

### Solutions

Ideally someone (not me) steps up and helps cffi build wheels for macos_arm64. No idea how hard this is.  

A workaround has been proposed:
```
pip install cffi --no-binary :all:
```

And a somewhat more complicated version:
```
LDFLAGS=-L$(brew --prefix libffi)/lib CFLAGS=-I$(brew --prefix libffi)/include pip install cffi --no-binary :all:
```

I've heard mixed results so far though.

It looks like installing via conda produces a working version.

If we can confirm that the above workaround works, we can add that in our installation instructions: "bla use conda or pip install cffi --no-binary :all:".

## 2. Callbacks are broken

Something to do with how callbacks work in cffi, and Mac nagging that the binary is unsigned. Read more here: https://cffi.readthedocs.io/en/latest/using.html#callbacks-old-style

The error that you'd get:

> MemoryError: Cannot allocate write+execute memory for ffi.callback(). You might be running on a system that prevents this. 

The cffi docs mention that the "new callback mechanism" does not have this problem. We only have 2 callbacks, so we might be able to fix this. I'll look into this.

cc @korijn @berendkleinhaneveld @marcdownie @SuperSimon81 @anentropic

**Comment by berendkleinhaneveld** [comment]
No luck... The weird thing is that I don't see my env var show up in the generated clang command:
```sh
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/opt/homebrew/Cellar/libffi/3.3_3/include -I/Users/cg/Library/Caches/pypoetry/virtualenvs/wgpu-5bSf_T1V-py3.9/include -I/Users/cg/.pyenv/versions/3.9.7/Library/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c c/_cffi_backend.c -o build/temp.macosx-11.6-arm64-3.9/c/_cffi_backend.o -iwithsysroot/usr/include/ffi
clang -bundle -undefined dynamic_lookup -L/opt/homebrew/opt/readline/lib -L/opt/homebrew/opt/readline/lib -L/Users/cg/.pyenv/versions/3.9.7/lib -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib -L/opt/homebrew/opt/readline/lib -L/opt/homebrew/opt/readline/lib -L/Users/cg/.pyenv/versions/3.9.7/lib -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib build/temp.macosx-11.6-arm64-3.9/c/_cffi_backend.o -L/opt/homebrew/Cellar/libffi/3.3_3/lib -lffi -o build/lib.macosx-11.6-arm64-3.9/_cffi_backend.cpython-39-darwin.so```

Build command I used was:
```sh
PKG_CONFIG_PATH=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/ffi poetry run pip install -vvv cffi --no-binary :all:
```
I've also tried to export the PKG_CONFIG_PATH var, but to no avail.

**Comment by berendkleinhaneveld** [comment]
> I noticed that there is a pre-release on pypi that may have "better" cffi wheels... can anyone try those?

Yes, it works! I also notice that arm wheels are available now for macOS, so no local builds are performed now. Shall I bump the version in `setup.py` or shall we wait for the official release? It seems that 1.15 also supports python 3.6 and up, like us.

**Comment by almarklein** [comment]
I did not expect the memoryerror to disappear so easily, but good news! :)

**Comment by almarklein** [comment]
Ok in #136 I read that its not crystal clear yet. I made #197 to track the memoryerror/callback. Closing this now.

**Comment by AdityaHPatwardhan** [comment]
I was facing a similar error while building cffi on MAC with M1 chip.
I was trying to install cffi using `pip3 install cffi` with an environment that used python 3.8
It was failing with `failed to build wheel` error.

But then I used environment that uses python 3.11 and somehow the issue got fixed. Just commenting a possible alternative here in case someone faces similar issue long after this issue is closed.
